### PR TITLE
Napatech counters v2.2

### PR DESCRIPTION
**Pull request by pjyoung1**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x ] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x ] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Adds packet counters when using Napatech hardware.
PR v2.2 addresses issues with previous PR.
-
-

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by pjyoung1** [comment]
What does the message  "continuous-integration/appveyor/pr — AppVeyor was unable to build non-mergeable pull request" mean?  Is this blocking the PR?

**Comment by inliniac** [comment]
Not sure what happened, haven't seen this message before. Can you resubmit? I can't seem to force a build.

File: src/runmode-napatech.c [hunk 1]
```diff
@@ -44,15 +46,23 @@
 
 static const char *default_mode = NULL;
 #ifdef HAVE_NAPATECH
-static int num_configured_streams = 0;
+
+#define MAX_STREAMS 256
+static uint16_t num_configured_streams = 0;
+
+uint16_t getNumConfiguredStreams(void) {
```
Comment (inliniac) [inline src/runmode-napatech.c]: style: start with G

File: src/runmode-napatech.c [hunk 2]
```diff
@@ -69,128 +79,108 @@ void RunModeNapatechRegister(void)
...
-    char live_dev_buf[9];
-    int use_all_streams;
-    ConfNode *ntstreams;
-    ConfNode *stream_id;
 
-    if (ConfGetBool("napatech.use-all-streams", &use_all_streams) == 0)
-    {
+    // Display the configuration mode
+    int use_all_streams;
+    if (ConfGetBool("napatech.use-all-streams", &use_all_streams) == 0) {
```
Comment (inliniac) [inline src/runmode-napatech.c]: this means that if the "napatech.use-all-streams" option is not in the yaml, suricata will exit. Is that intentional?

File: src/runmode-napatech.c [hunk 3]
```diff
@@ -200,16 +190,29 @@ static int NapatechInit(int runmode)
         exit(EXIT_FAILURE);
     }
 
-    switch(runmode) {
+    struct NapatechStreamDevConf *conf = SCMalloc(sizeof (struct NapatechStreamDevConf));
```
Comment (inliniac) [inline src/runmode-napatech.c]: I prefer SCCalloc as that initializes the memory to 0. Reduces chance of issues.

File: src/source-napatech.c [hunk 4]
```diff
@@ -149,81 +163,111 @@ void TmModuleNapatechDecodeRegister(void)
...
+    
+    NapatechThreadVars *ntv = SCMalloc(sizeof (NapatechThreadVars));
     if (unlikely(ntv == NULL)) {
         SCLogError(SC_ERR_MEM_ALLOC, "Failed to allocate memory for NAPATECH  thread vars.");
         exit(EXIT_FAILURE);
     }
 
     memset(ntv, 0, sizeof (NapatechThreadVars));
-    ntv->stream_id = stream_id;
+    ntv->stream_id = (uint64_t) stream_id;
```
Comment (inliniac) [inline src/source-napatech.c]: ntv->stream_id was changed to uint16_t earlier in this patch

File: src/source-napatech.c [hunk 5]
```diff
@@ -76,52 +72,63 @@ TmEcode NoNapatechSupportExit(ThreadVars *tv, const void *initdata, void **data)
...
-TmEcode NapatechStreamLoop(ThreadVars *tv, void *data, void *slot);
+TmEcode NapatechPacketLoopZC(ThreadVars *tv, void *data, void *slot);
 
 TmEcode NapatechDecodeThreadInit(ThreadVars *, const void *, void **);
 TmEcode NapatechDecodeThreadDeinit(ThreadVars *tv, void *data);
 TmEcode NapatechDecode(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueue *);
 
+// These are used as the threads are exiting to get a comprehensive count of
+// all the packets received and dropped.
+SC_ATOMIC_DECLARE(uint64_t, total_packets);
```
Comment (inliniac) [inline src/source-napatech.c]: I think you can also use the regular stats api here. Should be more efficient that using atomics.

File: src/util-napatech.c [hunk 6]
```diff
@@ -0,0 +1,755 @@
...
+
+static uint16_t TestStreamConfig(
+        NtInfoStream_t hInfo,
+        NtStatStream_t hStatStream,
+        NapatechStreamConfig stream_config[], 
+        uint16_t num_inst) 
+{
+    int status;
+    char buffer[80]; // Error buffer
+    NtStatistics_t stat; // Stat handle.
```
Comment (inliniac) [inline src/util-napatech.c]: if this gets reset in each loop iteration I prefer it to be declared in the loop

File: src/util-napatech.c [hunk 7]
```diff
@@ -0,0 +1,755 @@
...
+    // Read the info on this specific stream
+    info.cmd = NT_INFO_CMD_READ_STREAMID;
+    info.u.streamID.streamId = stream_id;
+    if ((status = NT_InfoRead(info_stream, &info)) != NT_SUCCESS) {
+        NT_ExplainError(status, buffer, sizeof (buffer) - 1);
+        SCLogError(SC_ERR_NAPATECH_INIT_FAILED, "NT_InfoRead() failed: %s\n", buffer);
+        exit(EXIT_FAILURE);
+    }
+
+    numaNode = info.u.streamID.data.numaNode;
```
Comment (inliniac) [inline src/util-napatech.c]: style: numa_node

File: src/util-napatech.c [hunk 8]
```diff
@@ -0,0 +1,755 @@
...
+            stream_config[inst].is_active = false;
+        }
+    }
+
+    return num_active;
+}
+
+
+
+static uint32_t updateStreamStats(ThreadVars *tv,
```
Comment (inliniac) [inline src/util-napatech.c]: style: Update*

File: src/util-napatech.c [hunk 9]
```diff
@@ -0,0 +1,755 @@
...
+
+
+        while(inst_id < num_streams) {
+            if (stream_config[inst_id].is_active) {
+                break;
+            } else {
+                ++inst_id;
+            }
+        }        
+        if (inst_id == num_streams) break;
```
Comment (inliniac) [inline src/util-napatech.c]: please use 2 lines as stepping through this code with gdb will be harder otherwise

File: src/util-napatech.c [hunk 10]
```diff
@@ -0,0 +1,755 @@
...
+        NT_ExplainError(status, error_buffer, sizeof (error_buffer) - 1);
+        SCLogError(SC_ERR_RUNMODE, "NT_InfoOpen() failed: %s\n", error_buffer);
+        return NULL;
+    }
+
+    if ((status = NT_StatOpen(&hStatStream, "StatsLoopStatsStream")) != NT_SUCCESS) {
+        // Get the status code as text
+        NT_ExplainError(status, error_buffer, sizeof (error_buffer));
+        SCLogError(SC_ERR_RUNMODE, "NT_StatOpen() failed: %s\n", error_buffer);
+        return NULL;
```
Comment (inliniac) [inline src/util-napatech.c]: should there be any cleanup wrt the NT_InfoOpen call before?

**Comment by pjyoung1** [comment]
Other than the  two exceptions I've noted above I've incorporated all of these into PR: Napatech counters v2.3

**Comment by pjyoung1** [comment]
My comments above don't seem to have shown up.  

The two outstanding issues are:
 
1. I'm using ATOMIC variables instead of the suricata counters to tally the totals that are output when the program is exited.  I understand the suggestion to use the suricata counters instead.  But, does it make sense for this very limited use-case?  

2.	The "napatech.use-all-streams" is as it was implemented originally by nPulse.  It may make sense to have this set to true by default.  But, an argument could also be made to require this to be explicitly set.

I'm more than happy to address these.  But, I will not be able to get to them for a day or two.  And we are concerned that we won't make the release.

**Comment by inliniac** [comment]
Replaced by #2844

**Pull request closed without merging**